trin-core/src/portalnet/find/query_tracer.rs

```diff
@@ -0,0 +1,144 @@
+}
+
+impl QueryTracer {
+    pub fn new(local_enr: &Enr) -> Self {
+        QueryTracer {
+            found_content_at: None,
+            origin: local_enr.to_base64(),
+            responses: HashMap::new(),
+            started_at: SystemTime::now(),
+        }
+    }
+
+    /// Records the receipt of a node as part of a query in the trace.
+    /// Adds a new `QueryResponse` object for `node` if it does not exist, updates if it does.
+    /// Timestamp for the object is only set on the first call for a given `node`.
+    pub fn node_responded_with(&mut self, enr: &Enr, responded_with: &Enr) {
+        let enr_string = enr.to_base64();
+        let responded_with_id = responded_with.to_base64();
+        self.responses
+            .entry(enr_string)
+            .or_insert_with(|| {
+                // Entry does not exist, create it and insert it.
+                let timestamp_u64 = QueryTracer::timestamp_ms_u64(self.started_at);
+                QueryResponse {
+                    timestamp_ms: timestamp_u64,
+                    responded_with: vec![],
+                }
+            })
+            .responded_with
+            .push(responded_with_id);
+    }
+
+    /// Mark the node as having responded, but provided no new route.
+    pub fn node_responded_without_progress(&mut self, enr: &Enr) {
+        let enr_string = enr.to_base64();
+        self.responses.entry(enr_string).or_insert_with(|| {
+            // Entry does not exist, create it and insert it.
+            let timestamp_u64 = QueryTracer::timestamp_ms_u64(self.started_at);
+            QueryResponse {
+                timestamp_ms: timestamp_u64,
+                responded_with: vec![],
+            }
+        });
+    }
+
+    /// Mark the node that responded with the content, and when it was received.
+    pub fn found_content_at(&mut self, enr: &Enr) {
+        let enr_string = enr.to_base64();
+        self.found_content_at = Some(enr_string.clone());
+        let timestamp_u64 = QueryTracer::timestamp_ms_u64(self.started_at);
+        self.responses.insert(
+            enr_string,
+            QueryResponse {
+                timestamp_ms: timestamp_u64,
+                responded_with: vec![],
+            },
+        );
+    }
+
+    /// Returns milliseconds since the time provided.
+    fn timestamp_ms_u64(since: SystemTime) -> u64 {
+        let timestamp_ms_u128 = SystemTime::now()
+            .duration_since(since)
+            .unwrap_or(Duration::default())
+            .as_millis();
+        // JSON serialization does not support u128. u64 can hold a few million years worth of milliseconds.
+        u64::try_from(timestamp_ms_u128).unwrap_or(u64::MAX)
+    }
+}
+
+/// Represents the response from a single node.
+#[derive(Debug, Serialize)]
+pub struct QueryResponse {
+    /// Milliseconds since query started.
+    pub timestamp_ms: u64,
+    /// Node IDs that this node responded with.
+    pub responded_with: Vec<ENRString>,
+}
```

Comment:
A few comments on this struct.

It makes more sense to me to use the absolute timestamp at the time the response was received. The latencies can be computed later.

A minor thing is to rename the field to `timestamp_millis` to be consistent with `std::time` naming. This would apply here and other places where `ms` is used.

I would propose that we rename `responded_with` to `enrs` to be consistent with the portal wire protocol naming.

Reply:
> It makes more sense to me to use the absolute timestamp

Any reason? Seems like it would just add a bit more data & visual chaos to the JSON output. Though I think I should add a `started_at` field in the output so that the absolute time of any given request can be calculated.

> A minor thing is to rename the field to timestamp_millis to be consistent with std::time naming. This would apply here and other places where ms is used.

Done 👍

> I would propose that we rename responded_with to enrs to be consistent with the portal wire protocol naming.

👍